compiler/rustc_passes/src/dead.rs

```diff
@@ -677,30 +677,91 @@ impl<'tcx> DeadVisitor<'tcx> {
-                    }
+                if let Some(def_id) = item_did.as_local()
+                    && self.live_symbols.contains(&def_id)
+                {
+                    return true;
                 }
             }
         }
         false
     }
 
-    fn warn_dead_code(
+    fn store_dead_field_or_variant(
         &mut self,
         id: hir::HirId,
         span: rustc_span::Span,
         name: Symbol,
+    ) {
+        let field_or_variant_local_def_id = |parent_node, node| -> Option<(LocalDefId, bool)> {
+            match (parent_node, node) {
+                (
+                    Node::Item(hir::Item {
+                        kind:
+                            hir::ItemKind::Struct(hir::VariantData::Struct(..), ..)
+                            | hir::ItemKind::Union(hir::VariantData::Struct(..), ..),
+                        def_id,
+                        ..
+                    }),
+                    Node::Field(..),
+                ) => Some((*def_id, true)),
+                (
+                    Node::Item(hir::Item { kind: hir::ItemKind::Enum(..), def_id, .. }),
+                    Node::Variant(..),
+                ) => Some((*def_id, false)),
+                (
+                    Node::Variant(hir::Variant { data: hir::VariantData::Struct(..), id, .. }, ..),
+                    Node::Field(..),
+                ) => self
+                    .tcx
+                    .hir()
+                    .find_parent_node(*id)
+                    .and_then(|id| self.tcx.hir().opt_local_def_id(id).map(|id| (id, true))),
+                _ => None,
+            }
+        };
+        if let Some(parent_hir_id) = self.tcx.hir().find_parent_node(id)
+            && let Some(parent_node) = self.tcx.hir().find(parent_hir_id)
+            && let Some(node) = self.tcx.hir().find(id)
+            && let Some((did, is_field)) = field_or_variant_local_def_id(parent_node, node)
+            && !name.as_str().starts_with('_')
+        {
+            if is_field {
+                self.dead_fields.entry(did).or_default().push((id, span, name));
+            } else {
+                self.dead_variants.entry(did).or_default().push((id, span, name));
+            }
+        }
+    }
+
+    fn warn_multiple_dead_codes(
+        &self,
+        dead_codes: &[(hir::HirId, Span, Symbol)],
         participle: &str,
     ) {
-        if !name.as_str().starts_with('_') {
-            self.tcx.struct_span_lint_hir(lint::builtin::DEAD_CODE, id, span, |lint| {
-                let def_id = self.tcx.hir().local_def_id(id);
-                let descr = self.tcx.def_kind(def_id).descr(def_id.to_def_id());
-                let mut err = lint.build(&format!("{} is never {}: `{}`", descr, participle, name));
-                let hir = self.tcx.hir();
-                if let Some(encl_scope) = hir.get_enclosing_scope(id)
+        if let Some((id, _, name)) = dead_codes.first()
+            && !name.as_str().starts_with('_')
```

Comment:
Why is this just checking `.first`?

Reply:
The first dead_code's `hir_id` is used as an argument of `struct_span_lint_hir` and `name` is used in an error message when `dead_codes` has a single element. Do you think this should check if all dead codes have the same `DefKind`?

Reply:
> What happens if one of the variants allows the lint?

Given the following code:
```rust
#[derive(Debug)]
pub struct Whatever {
    pub field0: (),
    field1: (), //~ERROR these fields are never read
    field2: (),
    #[allow(dead_code)]
    field3: (),
    field4: (),
}
```
The current output is:
```
warning: these fields are never read
 --> src/main.rs:4:5
  |
4 |     field1: (), //~ERROR these fields are never read
  |     ^^^^^^^^^^
5 |     field2: (),
  |     ^^^^^^^^^^
...
8 |     field4: (),
  |     ^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
  = note: `Whatever` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
```

Reply:
> What happens if the first variant allows the lint?

Given the following code:
```
#[derive(Debug)]
pub struct Whatever {
    #[allow(dead_code)]
    field1: (), //~ERROR these fields are never read
    field2: (),
    field3: (),
    field4: (),
}
```

The current output is:
```rust
warning: these fields are never read
 --> src/main.rs:5:5
  |
5 |     field2: (),
  |     ^^^^^^^^^^
6 |     field3: (),
  |     ^^^^^^^^^^
7 |     field4: (),
  |     ^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
  = note: `Whatever` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
```

Reply:
> What happens if the second variant forbids the lint?

However, the last case has the following problem, which I will correct later.

Given the following code:

```rust
#[derive(Debug)]
pub struct Whatever {
    field1: (), //~ERROR these fields are never read
    #[forbid(dead_code)]
    field2: (),
    field3: (),
    field4: (),
}
```

The current output is:

```
warning: these fields are never read
 --> src/main.rs:3:5
  |
3 |     field1: (), //~ERROR these fields are never read
  |     ^^^^^^^^^^
4 |     #[forbid(dead_code)]
5 |     field2: (),
  |     ^^^^^^^^^^
6 |     field3: (),
  |     ^^^^^^^^^^
7 |     field4: (),
  |     ^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default
  = note: `Whatever` has a derived impl for the trait `Debug`, but this is intentionally ignored during dead code analysis
```

Reply:
I fixed this problem in https://github.com/rust-lang/rust/pull/97853/commits/194df3ec8cce584903238f104a02d82ee076f91f.